README_INTERFACE_TESTS.md

```diff
@@ -0,0 +1,173 @@
+That will give you the right type hints as you're typing the code, and will also give you a little utility cli to verify that the tests you are writing are being correctly picked up by the same code that will eventually be used to run them.
+
+Throughout this example we will write some interface tests for the `ingress` interface.
+This is the expected behaviour for the interface:
+1) on `ingress-relation-joined`, the requirer should write its part of the data to the relation. 
+2) on `ingress-relation-changed`, if the provider has written its part of the data to the relation, then the provider should publish its part of the data; otherwise it should publish no data.
+
+## A minimal test case
+
+Specify the schema for the `ingress` interface databags in `./interfaces/ingress/v0/schema.py`.
+
+Create a python file at `<repo-root>/interfaces/ingress/v0/interface_tests/my_tests.py` with these contents:
+
+```python
+from scenario import State
+from interface_tester.interface_test import interface_test_case
+
+
+@interface_test_case(
+    event='ingress-relation-joined',
+    role='requirer',
+)
+def test_data_published_on_joined(output_state: State):
+    return
+```
+
+This is a complete interface test.
+
+## Validating the output state
+
+If the specification dictates something more than "the databag contents should match this schema in that event", for example:
+
+> on `ingress-relation-joined`, the requirer should write its part of the data to the relation and set its status to Active.
+
+The test would then become:
+
+```python
+from scenario import State
+from interface_tester.interface_test import interface_test_case
+
+
+@interface_test_case(
+    event='ingress-relation-joined',
+    role='requirer',
+)
+def test_data_published_on_joined(output_state: State):
+    assert output_state.status.unit.name == 'active'
+```
+
+
+## Testing a negative
+
+The second norm we have above contains a negative. We split it into two assertions to write the tests:
+
+> on `ingress-relation-changed`, if the requirer has provided 'data' and 'baz' information in its relation databags, then the provider should publish its part of the data.
+> on `ingress-relation-changed`, if the requirer has NOT provided 'data' and 'baz' information in its relation databags, then the provider should publish no data.
+
+This becomes two test cases:
+
+```python
+from scenario import State, Relation
+from interface_tester.interface_test import interface_test_case, SchemaConfig
+
+
+@interface_test_case(
+    event='ingress-relation-joined',
+    role='provider',
+    input_state=State(
+        relations=[Relation(
+            endpoint='foo',
+            interface='ingress',
+            remote_app_name='remote',  # this is our simulated requirer
+            remote_app_data={
+                'data': 'foo-bar',
+                'baz': 'qux'
+            }
+        )]
+    )
+)
+def test_data_published_on_joined_if_remote_has_sent_valid_data(output_state: State):
+    pass
```

Comment:
I know this is out of scope of this PR but the framework makes it non-trivial to check what is exactly validated. We're not asserting anything explicitly.

When I'm reviewing this test for example, I have a hard time figuring out, what exactly are we validating. The same goes for the next one where the devil's in the "schema=SchemaConfig.empty" detail.

Reply:
That's a good point, we can revisit the design of the tests later

Reply:
opened a new issue to track this